File: gen_file_with_specific_size/gen_file_with_specific_size.py
#!/usr/bin/env python
# coding: UTF-8

import os
import sys
import logging
import argparse
import random

import pdb
import pprint


def pretty_print(obj):
    pprint.pprint(obj)

def gen_hex():
    rndm = random.randrange(10**160)
    r = "%064x" % rndm
    return r[:128]

def gen_hex_chunk(size_bytes):
    """Generate a large chunk of hex data at once for better performance.
    
    Args:
        size_bytes: Size in bytes to generate (as hex characters)
    
    Returns:
        String of hex characters
    """
    # Each byte needs 2 hex characters, so we need size_bytes/2 random bytes
    num_random_bytes = (size_bytes + 1) // 2  # Round up for odd sizes
    # Generate random bytes more efficiently
    random_bytes = random.getrandbits(num_random_bytes * 8).to_bytes(num_random_bytes, byteorder='big')
    hex_str = random_bytes.hex()
    # Return exact size needed
    return hex_str[:size_bytes]

if __name__ == '__main__':
    parser = argparse.ArgumentParser(description = 'Generate file with specific size')
    parser.add_argument('-s', '--size',
                        action = 'store',
                        nargs = None,
                        const = None,
                        default = None,
                        type = int,
                        choices = None,
                        help = 'Please input filesize you want.',
                        metavar = None,
                        required = True)
    parser.add_argument('-d', '--directory-path',
                        action = 'store',
                        nargs = None,
                        const = None,
                        default = None,
                        type = str,
                        choices = None,
                        help = 'Please input absolute directory path.',
                        metavar = None,
                        required = True)
    parser.add_argument('-f', '--filename',
                        action = 'store',
                        nargs = None,
                        const = None,
                        default = None,
                        type = str,
                        choices = None,
                        help = 'Please input filename.',
                        metavar = None,
                        required = True)
    arguments = parser.parse_args()
    
    absolute_filepath = "%s/%s.txt" % (arguments.directory_path, arguments.filename)

    # 指定したファイルが既に存在したら例外で終了
    if os.path.exists(absolute_filepath):
        raise Exception('FileAlreadyExistsError')

    # 指定したファイルサイズが保存先のディスクの容量を上回っていれば例外で終了
    stats = os.statvfs(arguments.directory_path)
    available_gigabytes = stats.f_frsize * stats.f_bavail / 1024 / 1024 / 1024
    if available_gigabytes <= arguments.size:
        raise Exception('FilesizeExceedsTargetDriveError')

    target_size_bytes = arguments.size * 1024 * 1024 * 1024
    chunk_size_bytes = 10 * 1024 * 1024  # Write 10MB chunks at a time for better performance
    
    with open(absolute_filepath, 'w', buffering=8*1024*1024) as fs:  # 8MB buffer
        bytes_written = 0
        
        while bytes_written < target_size_bytes:
            # Calculate remaining size to avoid overshooting
            remaining_bytes = target_size_bytes - bytes_written
            
            # Determine how much to write this iteration
            write_size = min(chunk_size_bytes, remaining_bytes)
            
            # Generate and write hex data
            chunk = gen_hex_chunk(write_size)
            fs.write(chunk)
            bytes_written += len(chunk)
    
    sys.exit(0)

File: gen_file_with_specific_size/gen_file_with_specific_size.rb
require 'optparse'
require 'securerandom'
require 'sys-filesystem'
require 'pry'

OPTIONS = {}
OptionParser.new do |opt|
  opt.on('-s size', 'File size(GB)',                             Integer) {|v| OPTIONS[:size]           = v}
  opt.on('-d directory-path', 'Output Directory with Full path', String)  {|v| OPTIONS[:directory_path] = v}
  opt.on('-f filename', '--filename',                            String)  {|v| OPTIONS[:filename]       = v}
  opt.parse!(ARGV)
end

raise OptionParser::MissingArgument, 'FileSizeNotSprcified'      unless OPTIONS[:size]
raise OptionParser::MissingArgument, 'DirectoryPathNotSpecified' unless OPTIONS[:directory_path]
raise OptionParser::MissingArgument, 'FileNameNotSpecified'      unless OPTIONS[:filename]

stat = Sys::Filesystem.stat(OPTIONS[:directory_path])
available_gigabytes = stat.bytes_free / 1024 / 1024 / 1024

# 指定した保存ディレクトリのディスク容量が指定したファイルサイズ以下だった場合、例外
raise StandardError, 'FilesizeExceedsTargetDriveError' unless available_gigabytes > OPTIONS[:size]

absolute_file_path = "#{OPTIONS[:directory_path]}/#{OPTIONS[:filename]}.txt"

# 既にファイルが指定したディレクトリに存在する場合、例外とする
raise StandardError, 'FileAlreadyExistsError' if File.file?(absolute_file_path)

target_size_bytes = OPTIONS[:size] * 1024 * 1024 * 1024
chunk_size = 10 * 1024 * 1024  # Write 10MB chunks at a time for better performance

File.open(absolute_file_path, 'w') do |f|
  bytes_written = 0
  
  while bytes_written < target_size_bytes
    remaining_bytes = target_size_bytes - bytes_written
    
    # Adjust chunk size for the last write to avoid overshooting
    write_size = [chunk_size, remaining_bytes].min
    
    # Generate hex data efficiently
    # SecureRandom.hex(n) generates n random bytes and returns 2*n hex characters
    # So for write_size bytes, we need write_size/2 as the parameter
    hex_data = SecureRandom.hex(write_size / 2)
    
    # Truncate to exact size needed
    hex_data = hex_data[0, write_size] if hex_data.length > write_size
    
    f.write(hex_data)
    bytes_written += hex_data.length
  end
end

exit

